packages/flutter/lib/src/painting/debug_overflow_indicator.dart

```diff
@@ -0,0 +1,392 @@
+///     classes that use this indicator mixin.
+class DebugOverflowIndicatorMixin {
+  // This class is intended to be used as a mixin, and should not be
+  // extended directly.
+  factory DebugOverflowIndicatorMixin._() => null;
+
+  static const Color _black = const Color(0xBF000000);
+  static const Color _yellow = const Color(0xBFFFFF00);
+  static const double _kMarkerSize = 0.1;
+  static const double _debugMarkerFontSize = 7.5;
+  static const double _debugMarkerLabelPadding = 1.0;
+  static const TextStyle _debugMarkerTextStyle = const TextStyle(
+    color: const Color(0xFF900000),
+    fontSize: _debugMarkerFontSize,
+    fontWeight: FontWeight.w800,
+  );
+  static final Paint _debugMarkerPaint = new Paint()
+    ..shader = new ui.Gradient.linear(
+      const Offset(0.0, 0.0),
+      const Offset(10.0, 10.0),
+      <Color>[_black, _yellow, _yellow, _black],
+      <double>[0.25, 0.25, 0.75, 0.75],
+      TileMode.repeated,
+    );
+  static final Paint _debugLabelBackgroundPaint = new Paint()..color = const Color(0xFFFFFFFF);
+
+  final TextPainter _debugMarkerLabel = new TextPainter()
+    ..textDirection = TextDirection.ltr; // This label is in English.
+
+  TextDirection get textDirection => null;
+
+  RenderObject overflowRenderObject;
+
+  String get _containerName => runtimeType.toString();
+
+  /// Set to true to trigger a debug message in the console upon
+  /// the next paint call.
+  ///
+  /// Will be set to false after a call to
+  /// paint, so it must be reset each time a debug message is desired.
+  ///
+  /// Typically, this is set to true inside of the
+  /// [RenderObject.reassemble] override in the [RenderObject] that
+  /// owns this [DebugOverflowIndicatorMixin].
+  bool overflowReportNeeded = true;
+
+  /// An explanation as to what went wrong, and how to fix it.
+  ///
+  /// Override this to customize the hints given when an overflow occurs.  The
+  /// amount of the overflow in each direction, and the details of the
+  /// [RenderObject] that is overflowing will be printed along with this
+  /// message.
+  String get overflowHints {
+    return 'The edge of the $runtimeType that is overflowing has been '
+        'marked in the rendering with a yellow and black striped pattern. This '
+        'is usually caused by the contents being too big for the '
+        '$runtimeType.\n'
+        'This is considered an error condition because it indicates that there '
+        'is content that cannot be seen. If the content is legitimately bigger '
+        'than the available space, consider clipping it with a RectClip widget '
+        'before putting it in the $runtimeType, or using a scrollable '
+        'container, like a ListView, for example.';
+  }
+
+  /// Returns a [RelativeRect] that describes the overflow conditions
+  /// calculated.
+  ///
+  /// Values in the result will be positive if there is an overflow on
+  /// that side.
+  RelativeRect get overflow {
```

Comment:
consider making this `debugOverflow`, making it only return a non-null value in debug mode, and dropping all the caching logic, for greater simplicity.